### PR TITLE
Implement intrusive ptr

### PR DESCRIPTION
**Pull request by trivialfis**
Previously the PR is hosted on dmlc-core.  Let's merge it here for XGBoost first so we can get better perf for next release with JSON now being the default serialization format.

@hcho3

### CODE REVIEW

**Comment by hcho3** [comment]
@trivialfis Do you have some benchmark results?

**Comment by trivialfis** [comment]
The model is trained with covtype dataset, generated 21000 trees in total with
max_depth set to 6.  The size of saved JSON file is about 96.1MB.

|                                        | Load               | Save               |
|:---------------------------------------|:-------------------|:-------------------|
| Before adding categorical data support | 4.2141406536102295 | 3.429635524749756  |
| Master                                 | 4.652015209197998  | 3.8093554973602295 |
| With new categorical split format      | 4.365228891372681  | 3.491779327392578  |
| With intrusive ptr                     | 2.460336446762085  | 2.5140161514282227 |

File: include/xgboost/intrusive_ptr.h [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+#ifndef XGBOOST_INTRUSIVE_PTR_H_
+#define XGBOOST_INTRUSIVE_PTR_H_
+
+#include <atomic>
+#include <cinttypes>
+#include <functional>
+
+namespace xgboost {
+/*!
+ * \brief Helper class for embedding reference counting into client objects.
```
Comment (hcho3) [inline include/xgboost/intrusive_ptr.h:15]: We should document why it's okay to use `std:: memory_order_relaxed` when incrementing the reference count.
```suggestion
 * \brief Helper class for embedding reference counting into client objects.
 * See https://www.boost.org/doc/libs/1_74_0/doc/html/atomic/usage_examples.html
 * for discussions of memory order.
```

File: include/xgboost/intrusive_ptr.h [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  int32_t Count() const { return count_.load(std::memory_order_relaxed); }
+};
+
+/*!
+ * \brief User defined function for returing embedded reference count.
+ */
+template <typename T> IntrusivePtrCell &IntrusivePtrRefCount(T const *ptr) noexcept;
+
+/*!
+ * \brief Implementation of Intrusive Pointer.
```
Comment (hcho3) [inline include/xgboost/intrusive_ptr.h:41]:
```suggestion
 * \brief Implementation of Intrusive Pointer, a smart pointer that points to an object
 *        with an embedded reference counter. The underlying object must implement
 *        a friend function IntrusivePtrRefCount() that returns the ref counter
 *        (of type IntrusivePtrCell). Embedding the ref counter in the managed object
 *        improves performance, compared to allocating a separate ref counter.
```

**Review by hcho3 — APPROVED** [review]

File: include/xgboost/intrusive_ptr.h (hunk 2 above)
Comment (trivialfis) [reply include/xgboost/intrusive_ptr.h:41]: Let me add an example usage.

**Pull request merged**